File: output/xCodeEval_8000/C++/C/7b2ed196ac86d9c6a19962e13d67ff47.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

#define max(a, b) ((a) > (b) ? (a) : (b))

typedef long long ll;

ll sum(int *arr, int size) {
    ll s = 0;
    for (int i = 0; i < size; i++) {
        s += arr[i];
    }
    return s;
}

void solve() {
    int n, k;
    scanf("%d %d", &n, &k);

    int *sums = (int *)malloc(n * sizeof(int));
    int **arrs = (int **)malloc(n * sizeof(int *));
    for (int i = 0; i < n; i++) {
        int t;
        scanf("%d", &t);
        arrs[i] = (int *)malloc(t * sizeof(int));
        for (int j = 0; j < t; j++) {
            scanf("%d", &arrs[i][j]);
        }
        if (t > k) {
            arrs[i] = realloc(arrs[i], k * sizeof(int));
        }
        sums[i] = sum(arrs[i], t);
    }

    ll **dps = (ll **)malloc(13 * sizeof(ll *));
    for (int i = 0; i < 13; i++) {
        dps[i] = (ll *)malloc((k + 1) * sizeof(ll));
        memset(dps[i], 0, (k + 1) * sizeof(ll));
    }

    for (int i = 0; i < n; i++) {
        dps[0][k] = max(dps[0][k], dps[0][k]);
        ll v = sums[i];
        int k1 = arrs[i].size();
        for (int j = k; j >= k1; j--) {
            dps[0][j] = max(dps[0][j], dps[0][j - k1] + v);
        }
    }

    for (int l = 0; l < n; l++) {
        for (int r = l + 1; r < n; r++) {
            for (int i = 1; i <= 12; i++) {
                for (int j = k; j >= 0; j--) {
                    dps[i][j] = max(dps[i][j], dps[i - 1][j]);
                    ll sum = 0;
                    for (int x = 0; x < min(k, arrs[l].size()); x++) {
                        sum += arrs[l][x];
                        dps[i][j] = max(dps[i][j], dps[i - 1][j - x - 1] + sum);
                    }
                }
            }
        }
    }

    printf("%lld\n", dps[12][k]);

    for (int i = 0; i < n; i++) {
        free(arrs[i]);
    }
    free(arrs);
    free(sums);
    for (int i = 0; i < 13; i++) {
        free(dps[i]);
    }
    free(dps);
}

int main() {
    solve();
    return 0;
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Go/7b2ed196ac86d9c6a19962e13d67ff47.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"sort"
	"strconv"
)

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func sum(arr []int) int {
	total := 0
	for _, num := range arr {
		total += num
	}
	return total
}

func rec(l, r, k int, dps [][]int, sums []int, arrs [][]int) int {
	ans := 0
	carr := arrs[l]
	dp := dps[k]
	if l+1 == r {
		ans = max(ans, dp[k])
		sum := 0
		for i := 0; i < min(k, len(carr)); i++ {
			sum += carr[i]
			ans = max(ans, dp[k-i-1]+sum)
		}
	} else {
		m := (l + r) / 2
		dps[dpi+1] = make([]int, k+1)
		copy(dps[dpi+1], dps[dpi])
		for i := l; i < m-1; i++ {
			v := sums[i]
			k1 := len(arrs[i])
			for j := k; j >= k1; j-- {
				dp[j] = max(dp[j], dp[j-k1]+v)
			}
		}
		ans = max(ans, rec(m, r, k, dps, sums, arrs))
		copy(dps[dpi+1], dps[dpi])
		for i := m; i < r-1; i++ {
			v := sums[i]
			k1 := len(arrs[i])
			for j := k; j >= k1; j-- {
				dp[j] = max(dp[j], dp[j-k1]+v)
			}
		}
		ans = max(ans, rec(l, m, k, dps, sums, arrs))
	}
	return ans
}

func DSum(reader io.Reader, writer io.Writer) {
	scanner := bufio.NewScanner(reader)
	scanner.Split(bufio.ScanWords)

	// Read n and k
	scanner.Scan()
	n, _ := strconv.Atoi(scanner.Text())
	scanner.Scan()
	k, _ := strconv.Atoi(scanner.Text())

	sums := make([]int, n)
	arrs := make([][]int, n)
	for i := 0; i < n; i++ {
		// Read t
		scanner.Scan()
		t, _ := strconv.Atoi(scanner.Text())

		// Read arr
		arr := make([]int, t)
		for j := 0; j < t; j++ {
			scanner.Scan()
			arr[j], _ = strconv.Atoi(scanner.Text())
		}

		// Sort arr in descending order
		sort.Sort(sort.Reverse(sort.IntSlice(arr)))

		// Truncate arr to length k
		if t > k {
			arr = arr[:k]
		}

		// Calculate sum of arr
		sums[i] = sum(arr)

		// Store arr
		arrs[i] = arr
	}

	// Initialize DPS array
	dps := make([][]int, 13)
	for i := 0; i < 13; i++ {
		dps[i] = make([]int, k+1)
	}

	// Call recursive function to solve the problem
	ans := rec(0, n, k, dps, sums, arrs)

	// Write the answer to the output
	fmt.Fprintln(writer, ans)
}

func main() {
	DSum(os.Stdin, os.Stdout)
}
